### DISCUSSION TITLE
export CSV files with XML code inside

### DISCUSSION THREAD

**Discussion by pr-apes** [speedata/publisher#408] [Q&A] [answered]
@pgundlach,

I would like to be able to include some XML code in a CSV file, such as in:

``` xml
"John","This was his page: <URL>https://domain.com</URL>, but that was years ago."
```

But when I export it to XML, the XML code from CSV is escaped.

Is there a way to avoid it, or even better that the layout recognizes it as a native element?

Many thanks for your help.

**Comment by pgundlach** [comment]
You need to parse the CSV output and adjust the code yourself.

**Reply by pr-apes** [reply]
@pgundlach,

many thanks for your reply.

I see that I have to parse the XML tags in XML data.

Sorry, but what I don't see is how having the following CSV input:

``` xml
"abc","x<em>y</em>z"
````

the encoding into XML is:

``` xml
<data><row><cell>abc</cell><cell>x&lt;em&gt;y&lt;/em&gt;z</cell></row>
```

This makes impossible to add format inside the text in CSV (among other things).

Would it be possible to have an option in `xml.encode_table()` (such as `raw` or something similar), that doesn't escape XML?

Many thanks for your help.

**Comment by pgundlach** [comment, marked as answer]
It would be possible, but currently not on my priority list to implement.  Everything like what you want is possible without a "raw" option.

You know that you can use `sd:decode-html()` for some of these use cases?

**Reply by pr-apes** [reply]
@pgundlach,

many thanks for your reply. I think `sd:decode-html()` answers to my question.

I knew `sd:decode-html(<Node>)`, but I thought that this was related to handle `<i>italics</i>` as it were `<I>italics</I>` in _Publisher_.

Checking the documentation, it reads in English:

> sd:decode-html()
>
> : Change text such as `<i>italic</i>` into HTML markup.

The original version is clear even if you cannot read German (forgotten backticks):

> sd:decode-html()
>
> : Wandelt Texte wie `&lt;i&gt;Kursiv&lt;/i&gt;` in entsprechendes HTML-Markup.

Now that this is clear, my question is: does `sd:decode-html()` decode XML character entities (such as the ones contained in https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references)?

I wonder whether it would make sense to rename `sd:decode-html()` with `sd:decode-char-entities()` (or something similar).

I have no idea on how widely the function is used, so breaking backwards compatibility might be an issue too.

Many thanks for your help.

**Reply by pgundlach** [reply]
> Now that this is clear, my question is: does `sd:decode-html()` decode XML character entities (such as the ones contained in https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references)?
> 
> I wonder whether it would make sense to rename `sd:decode-html()` with `sd:decode-char-entities()` (or something similar).

you can always try ;-) 

```xml
<data>
    &amp;#x002A;
</data>
```
and 

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
    xmlns:sd="urn:speedata:2009/publisher/functions/en">

    <Record element="data">
        <PlaceObject>
            <Textblock>
                <Paragraph>
                    <Value select="sd:decode-html(.)"></Value>
                </Paragraph>
            </Textblock>
        </PlaceObject>
    </Record>
</Layout>
```

gives a * which is what I'd expect. Or am I missing something here?

**Reply by pgundlach** [reply]
> > sd:decode-html()
> > : Change text such as `<i>italic</i>` into HTML markup.
> 
> The original version is clear even if you cannot read German (forgotten backticks):
> 
> > sd:decode-html()
> > : Wandelt Texte wie `&lt;i&gt;Kursiv&lt;/i&gt;` in entsprechendes HTML-Markup.

thank you for spotting the malformed markup. I've fixed this in the English translation of the manual - will be online for the next version which will take some time (vacation)

**Reply by pr-apes** [reply]
If you are on holidays, please enjoy your time.

Forget everything from this repository.

And please, say hi when you come back.

Again, happy holidays, have fun and enjoy your relax.
